dns.c

```diff
@@ -47,18 +51,38 @@ int type;
  errno = 0;
  if (!stralloc_copy(&glue,domain)) return DNS_MEM;
  if (!stralloc_0(&glue)) return DNS_MEM;
- responselen = lookup(glue.s,C_IN,type,response.buf,sizeof(response));
+ if (!responsebuflen) {
+  if ((response.buf = malloc(PACKETSZ+1)))
+   responsebuflen = PACKETSZ+1;
+  else return DNS_MEM;
```

Comment:
[#check-err-success]
This switches from checking for failure (`if (!(buf = malloc(4)) ERR;`) to checking for success, which conveniently looks like `try { ... } catch { ... }`, but adds more indentation depth, and makes switching just in one place could make failure handling more tedious to check through the source.

Reply:
I generally prefer guard clauses too, for the reasons you state. The original patch gives us the code in this form, though, and although this particular clause would be simple enough to rephrase, I'm not generally feeling confident enough in my understanding of this code to refactor and re-test it. Mostly what I know about it is that I've been running it in production for a long time without apparent problems.

I'd be happy to see a followup PR from you or @leahneukirchen or @DerDakon or anyone else who feels more confident doing these cleanups :-)